File: backend/app/schemas/animal.py
from pydantic import BaseModel, Field
from typing import Optional
from datetime import date, datetime
from enum import Enum

class EspecieOut(BaseModel):
    especie_id: int
    nombre_comun: str

    class Config:
        from_attributes = True

class RazaOut(BaseModel):
    raza_id: int
    nombre_raza: str

    class Config:
        from_attributes = True

class AnimalSimpleOut(BaseModel):
    animal_id: int
    numero_trazabilidad: str
    nombre_identificatorio: Optional[str] = None

    class Config:
        from_attributes = True

class EstadoAnimal(str, Enum):
    Activo = "Activo"
    Vendido = "Vendido"
    Muerto = "Muerto"
    Descartado = "Descartado"

class AnimalOut(BaseModel):
    animal_id: int
    numero_trazabilidad: str
    nombre_identificatorio: Optional[str] = None
    especie: EspecieOut
    raza: RazaOut
    sexo: str
    fecha_nacimiento: Optional[date] = None
    madre: Optional[AnimalSimpleOut] = None
    padre: Optional[AnimalSimpleOut] = None
    estado_actual: EstadoAnimal
    fecha_registro: datetime
    observaciones_generales: Optional[str] = None

    class Config:
        from_attributes = True

class AnimalBase(BaseModel):
    numero_trazabilidad: str = Field(..., max_length=50)
    nombre_identificatorio: Optional[str] = Field(None, max_length=100)
    especie_id: int
    raza_id: int
    sexo: str = Field(..., pattern="^[MH]$")
    fecha_nacimiento: Optional[date] = None
    madre_id: Optional[int] = None
    padre_id: Optional[int] = None
    estado_actual: EstadoAnimal = EstadoAnimal.Activo
    observaciones_generales: Optional[str] = None

class AnimalCreate(AnimalBase):
    pass

class AnimalUpdate(AnimalBase):
    pass

File: types/animal.ts
// types/animal.ts

export interface EspecieOut {
  especie_id: number;
  nombre_comun: string;
}

export interface RazaOut {
  raza_id: number;
  nombre_raza: string;
}

export interface AnimalSimpleOut {
  animal_id: number;
  numero_trazabilidad: string;
  nombre_identificatorio?: string;
}

export enum EstadoAnimal {
  Activo = "Activo",
  Vendido = "Vendido",
  Muerto = "Muerto",
  Descartado = "Descartado",
}

export interface AnimalOut {
  animal_id: number;
  numero_trazabilidad: string;
  nombre_identificatorio?: string;
  especie: EspecieOut;
  raza: RazaOut;
  sexo: string;
  fecha_nacimiento?: string; // Fecha en formato string (ISO 8601)
  madre?: AnimalSimpleOut;
  padre?: AnimalSimpleOut;
  estado_actual: EstadoAnimal;
  fecha_registro: string; // Fecha y hora en formato string (ISO 8601)
  observaciones_generales?: string;
}

export interface AnimalBase {
  numero_trazabilidad: string;
  nombre_identificatorio?: string;
  especie_id: number;
  raza_id: number;
  sexo: string;
  fecha_nacimiento?: string;
  madre_id?: number;
  padre_id?: number;
  estado_actual: EstadoAnimal;
  observaciones_generales?: string;
}

export interface AnimalCreate extends AnimalBase {}

export interface AnimalUpdate extends AnimalBase {}
